### PR TITLE
Added Product Mass Error to Output, cleaned code

### PR DESCRIPTION
**Pull request by zrolfs**

### CODE REVIEW

File: EngineLayer/MatchedIonMassesListOnlyMatches.cs [hunk 1]
```diff
@@ -47,18 +46,12 @@ IEnumerator IEnumerable.GetEnumerator()
             return matchedIonDictPositiveIsMatch.GetEnumerator();
         }
 
-        public override bool Equals(object obj)
```
Comment (zrolfs) [inline EngineLayer/MatchedIonMassesListOnlyMatches.cs]: Not sure if this should be deleted or not; was removed as part of merge conflict.

File: EngineLayer/MetaMorpheusEngine.cs [hunk 2]
```diff
@@ -41,98 +41,161 @@ protected MetaMorpheusEngine(List<string> nestedIds)
...
+                currentTheoreticalMass = sortedTheoreticalProductMassesForThisPeptide[currentTheoreticalIndex];
+            } while (double.IsNaN(currentTheoreticalMass) && currentTheoreticalIndex < sortedTheoreticalProductMassesForThisPeptide.Length - 1);
 
             if (double.IsNaN(currentTheoreticalMass))
-                return 0;
+                return;
 
             double currentTheoreticalMz = currentTheoreticalMass + Constants.protonMass;
 
-            int testTheoreticalIndex;
```
Comment (zrolfs) [inline EngineLayer/MetaMorpheusEngine.cs]: flipped order here. Used to skip theoreticals until one was close to an experimental, but experimentals vastly outnumber theoreticals, so this speeds up the loop.

File: EngineLayer/MetaMorpheusEngine.cs [hunk 3]
```diff
@@ -185,11 +248,6 @@ private static void AddComplementaryPeaks(ref double[] experimental_mzs, ref dou
         {
             List<MzPeak> complementaryPeaks = new List<MzPeak>();
 
-            //keep original peeks
```
Comment (zrolfs) [inline EngineLayer/MetaMorpheusEngine.cs]: No longer keep original peaks, so one search is conducted with original, and one with only comp ions.

File: EngineLayer/ModernSearch/ModernSearchEngine.cs [hunk 4]
```diff
@@ -157,58 +157,49 @@ protected override MetaMorpheusEngineResults RunSpecific()
...
-                    var theAdd = 1 + spectrum.MassSpectrum[i].Intensity / spectrum.TotalIonCurrent;
-                    var experimentalPeakInDaltons = spectrum.MassSpectrum[i].Mz - Constants.protonMass;
-                    GeneratePeptideScores(theAdd, experimentalPeakInDaltons, peptideScores);
-                }
+                var theAdd = 1 + spectrum.MassSpectrum[i].Intensity / spectrum.TotalIonCurrent;
+                var experimentalPeakInDaltons = spectrum.MassSpectrum[i].Mz - Constants.protonMass;
+                GeneratePeptideScores(theAdd, experimentalPeakInDaltons, peptideScores, listSeenipos);
             }
-            else
+            listSeenipos.Clear();
```
Comment (zrolfs) [inline EngineLayer/ModernSearch/ModernSearchEngine.cs]: Allows double peak counting if one is seen in original spectra and one in complementary ions

File: EngineLayer/NonSpecificEnzymeSearch/NonSpecificEnzymeEngine.cs [hunk 5]
```diff
@@ -212,90 +212,7 @@ private double Accepts(double scanPrecursorMass, CompactPeptide peptide, Terminu
                 }
             }
             return 0;
-        }
-
-        private void CalculatePeptideScores(IMsDataScan<IMzSpectrum<IMzPeak>> spectrum, double[] peptideScores, double thePrecursorMass)
```
Comment (zrolfs) [inline EngineLayer/NonSpecificEnzymeSearch/NonSpecificEnzymeEngine.cs]: Refactoring has made these methods redundant.

File: Test/AddCompIonsTest.cs [hunk 6]
```diff
@@ -63,7 +63,7 @@ public static void TestAddCompIonsClassic()
 
             Assert.AreEqual(allPsmsArray[0].ScanNumber, allPsmsArray2[0].ScanNumber);
 
-            Assert.IsTrue(allPsmsArray2[0].Score < allPsmsArray[0].Score * 2 && allPsmsArray2[0].Score + 2 > allPsmsArray[0].Score * 2);
+            Assert.IsTrue(allPsmsArray2[0].Score < allPsmsArray[0].Score *3 && allPsmsArray2[0].Score + 1 > allPsmsArray[0].Score *3);
```
Comment (zrolfs) [inline Test/AddCompIonsTest.cs]: Update unit test, as previously peaks were not being double counted in classic

File: Test/AddCompIonsTest.cs [hunk 7]
```diff
@@ -139,7 +139,7 @@ public static void TestCompIons_ModernSearch()
 
             Assert.AreEqual(allPsmsArray[0].ScanNumber, allPsmsArray2[0].ScanNumber);
 
-            Assert.IsTrue(allPsmsArray2[0].Score < allPsmsArray[0].Score * 2 && allPsmsArray2[0].Score + 2 > allPsmsArray[0].Score * 2);
+            Assert.IsTrue(allPsmsArray2[0].Score < allPsmsArray[0].Score * 2 && allPsmsArray2[0].Score > allPsmsArray[0].Score + 3);
```
Comment (zrolfs) [inline Test/AddCompIonsTest.cs]: Recalculation of score bumped a decimal over the tolerance, but I think it's a better way to do it, so updated unit test

File: EngineLayer/MatchedIonMassesListOnlyMatches.cs (hunk 1 above)
Comment (stefanks) [reply EngineLayer/MatchedIonMassesListOnlyMatches.cs]: Please do not remove this. Restore if was deleted.

File: EngineLayer/MetaMorpheusEngine.cs (hunk 2 above)
Comment (stefanks) [reply EngineLayer/MetaMorpheusEngine.cs]: Please run a 3-task (classic search+gptmd+classic search) procedure on 3 different spectra files, including both qhf and velos files. Both before and after this change, but run on the same computer. Maybe Artemis? Then please give me the AllResults.txt for all six runs. 

This is a fundamental change to the search procedure, and I want to make sure it didn't make things slower.

File: EngineLayer/MetaMorpheusEngine.cs (hunk 3 above)
Comment (stefanks) [reply EngineLayer/MetaMorpheusEngine.cs]: Oooh! Clever

File: EngineLayer/MetaMorpheusEngine.cs (hunk 2 above)
Comment (zrolfs) [reply EngineLayer/MetaMorpheusEngine.cs]: Will do.

File: EngineLayer/ModernSearch/ModernSearchEngine.cs (hunk 4 above)
Comment (stefanks) [reply EngineLayer/ModernSearch/ModernSearchEngine.cs]: please move the clear to the if statement below

File: EngineLayer/MetaMorpheusEngine.cs [hunk 8]
```diff
@@ -41,98 +41,180 @@ protected MetaMorpheusEngine(List<string> nestedIds)
...
             int MatchingProductsHere = 0;
             double MatchingIntensityHere = 0;
 
             // speed optimizations
             double[] experimental_mzs = thisScan.MassSpectrum.XArray;
             double[] experimental_intensities = thisScan.MassSpectrum.YArray;
 
-            if (addComp)
+            if(addComp)
                 AddComplementaryPeaks(ref experimental_mzs, ref experimental_intensities, precursorMass, lp);
```
Comment (stefanks) [inline EngineLayer/MetaMorpheusEngine.cs]: rename this method

**Pull request merged**